### DISCUSSION TITLE
Difficulties with form request handling in a simple CRUD-application

### DISCUSSION THREAD

**Discussion by jrathert** [symfony/symfony#49857] [Q&A] [answered]
Hi,

I am quite new to Symfony (so this is a BEGINNER question) and are just trying to implement a VERY simple web application: 

- When a user accesses the page (without and request parameter), a UUID is generated and the  user should enter her name, email and shirt color.
- If she submits, data (uuid, name, email, shirt color) is stored in a database and the user gets an email with a link (leveraging the uuid) that she can use to come back later and change her mind/data (of the shirt color, or even her name or email).

That's all. Basically a CRUD without the "D", and yes, not super secure, but this is what we need.

I thought this would be super easy with Symfony, leveraging doctrine and forms. Any yes, there are many good examples out there that describe how to easily build such an application, simply creating an entity (more or less a POPO with some validation rules), a respective form and type and a controller and then let Symfony do the work. Even the official documentation shws this with a "Task" object. Cool, 10 min work (at most) which I think is great for such a simple use case. Validation of new data as well as then saving new users is super easy.

But the challenge I face: Whenever I want to _change_ existing data, i.e. display the user's data and let her update it, she might do the mistake and delete some of the data, say her name. While in case of a _new_ entry, the form validation would then show a user friendly error in that case (e.g., "Name should not be blank"), in case of an update, Symfony throws an exception, as a null value cannot be copied to what is expected to be a string.

```
// ...
#[Route('/edit/{id}', name: 'app_save')]
public function edit(Request $request, EntityManagerInterface $entityManager, User $user): Response
{
    $form = $this->createForm(UserType::class, $user);
    $form->handleRequest($request);      // <----- Exception thrown if one of the text fields is empty
    if ($form->isSubmitted() && $form->isValid()) {
// ...

```

I searched for that and found out that it is exactly what has been described here: https://github.com/symfony/symfony/issues/39106

It seems one should use DTOs and with that decouple the "business entity" (which should always be "valid") from form data (which may not). I really did a lot of searching, reading docs, even watching YT videos etc. - I simply cannot find a SIMPLE example on how this should work - the "update case" is never shown, only the "create" where you simply copy all form->data(). So I did it by myself, ending up with two more or less similar objects, one being the "entity" that is stored to/retrieved from the database with a couple of constraints (Non-Null, etc.) and one being the DTO with more or less the same data and the same constraints just used in the form (but despite the constraints technically allowing null properties!). And of course I wrote a lot of stupid code to copy data between the DTO and the business entity when preparing/processing the form.

**Question:** Did I make a fundamental mistake in creating all the stupid DTO / copy code? If yes, can someone please help me finding a minimal example that avoids that and does validation properly for "null"-values in the update case?

**Comment:** If I did NOT do a mistake and that is in general how it should work: I would say it is perfectly fine for more complex situations where the business entity really differs from the form data, e.g., because there is some more complex logic to derive the entity from the data etc. BUT: I strongly believe Symfony should provide an easy way for the simple "CRUD"-type of application - and as far as I can say, the "U" part is really annoying. Maybe one should have some sort of configuration value that makes the request handling in the form do it in a bit of a "relaxed" way (what obviously is the case when adding new data)?  

Thanks for any help!

  Jonas

**Comment by bhagman** [comment]
Fellow Symfony beginner here.

I followed [this Symfony tutorial on YT](https://youtu.be/QPky3r2prEI?t=12667) and it has some great tips to get started.

If I understand your problem correctly, I believe your issue is coming from the validation at the ORM level.  AFAIU the Form Type will do further validation at the ORM level if you've defined the 'data_class' in the `configureOptions()` for your Form Type.

```php
class YourFormType extends AbstractType
{
    // ...

     public function configureOptions(OptionsResolver $resolver): void
    {
        $resolver->setDefaults([
            'data_class' => YourDataEntity::class,
        ]);
    }
}
```

To clarify, in your Entity class, each attribute you want ensure is not blank should have the `#[Assert\NotBlank()]` decorator.  This will ensure the field is not null (or blank) in `handleRequest()`.

```php
#[ORM\Entity(repositoryClass: YourDataRepository::class)]
class YourData
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column]
    private ?int $id = null;

    #[ORM\Column(length: 255)]
    #[Assert\NotBlank()]  // <----
    #[Assert\Length(min: 3,max: 255)]
    private ?string $name = null;

    /// ...
}
```

**Comment by jrathert** [comment]
Hi, 

thanks for your answer. Unfortunately, it does not really help. In general, I understand validation and how it works with the Assert decorators. This is not my problem, my code basically looks like your examples.

The problem I have is in the "edit-case": My entity has all values set as expected and the form shows all the values. The user now deletes one of the mandatory fields. I would expect validation to give an error and the form to be redisplayed showing the error (something klik "field should not be empty". But: ` form->handleRequest($request)` simply throws an exception if the user deletes some existing value that should not be null/blank, as the request then delivers a null value and as the entity does not allow this null value in its setter. (It is NOT thrown when creating an incomplete new value, as the entity is actually initialized using a null value per default, so there is no setter involved or whatsoever).

You can read many (probably beginner) people struggling with that. One is the github issue I linked to above (https://github.com/symfony/symfony/issues/39106 - EXACTLY my problem), another example can be found here: https://stackoverflow.com/questions/67772645/symfony-5-forms-and-entities-notblank-constraint-with-not-nullable-string-en and there are some others I do not have at hand right now. 

It seems a bit weird that a framework such as symfony that should make life simpler makes this so hard to implement and does not provide a good simple example. The "solutions" I found where:

- Changing the entity class generated by `make:entity` to allow null values in setter - allowing the entity to have an "invalid" state (which it might have anyway after an incomplete initialization, as all fields are initialized to be null...)
- Implementing a DTO, with less restrictions (i.e. nullable fields) - requiring a lot of "almost double code" in case of a simple CRUD application
-  Asking the form to replace any "null" value from the request by an empty string in the form builder (via `->add('fieldName', TextType::class, ['empty_data' => '']) `) - quite a bit of effort for such a simple thing

I really wonder why there is no simple configuration value for a form to handle "null" values in requests in such simple cases... While I find many statements that "it should work that way", I do not find any good explanation/rationale.

But as said, this all is a beginners POV.

**Reply by ahundiak** [reply]
Why do you feel that setting empty_data involves quite a bit of effort?  

The problem is that the way http form values are posted means you cannot distinguish between null and an empty string.  Pretty much just the way things are.  You can add a no blank validation directly in your form's html that will catch things before posting.  You can add a bit of javascript to do the same. And while I have never done it for this particular requirement you could probably tweak Symfony's property accessor to convert null strings to empty strings.  Need to be careful because you may find out that you want some strings to be nullable in your model.

As far as the notion that an entity should always have a valid state goes.  The very fact that the form component only sets one value at a time means the entity will almost always be invalid until the data transfer is complete.  So you are always violating the concept anyways.

**Reply by jrathert** [reply]
Hi,

thanks for your response! An yes, you are right, setting `empty_data` is not that much of an effort for a small entity (esp. compared to the DTO + transformation code etc.). And yes, the "objects" transferred by html forms can probably never considered "valid". And yes, in many scenarios, entities are more complex than the one I have in my super-simple CRUD application and might actually require nullable strings. So all valid points.

Maybe I am just puzzled by the fact that so many things are made _simple_ in Symfony - you get make-tools for entities, database migration scripts etc. All examples show how these created entities can be "injected" into the database easily and used with little to no additional code in a simple app where you directly use them with the form: you simply create the object (with `new`), transfer the form-data into it (with `form->handleRequest()`) and then store it in the database. So you think the editing case works exactly the same. As it is not shown in any example or documentation you just try it the same way, but all you get is the exception from `form->handleRequest()` - which is super-confusing as it behaves as expected in the case where you create an entity. A simple "hey, there is a trap, and here is how you solve it, use option a), b) or c), depending on your needs" would be helpful. (If I were a bit more experienced I'd volunteer to write that part of the documentation, but I simply do not know enough about Symfony yet.)

Anyway, I will use the `empty_data`-approach as it works (best) in my case.

**Reply by ahundiak** [reply]
Just out of curiosity have you experimented with the `make:crud` command?  I don't use it myself but I'd expect the generated forms would handle this sort of thing.

**Reply by jrathert** [reply]
Thanks for pointing to it, will have a look at it. (TBH: I didn't even know it exists, my learning curve is still evolving ;-).

BTW: I found some old, but maybe relevant video: https://vimeo.com/134178140#t=12m31s Marco from the Doctrine project talks about entity validity and mentions that Smyfony forms and the documented examples (with forms accessing entities) are probably not the best thing to do (he uses the term "terrible"). The talk is from 2015, but apparently, the situation has not improved since then.

**Reply by jrathert** [reply]
Adding to the above: I tried with `make:crud` and it basically has the same problem. It is a bit invisible, though, as out-of-the-box it does client side validation, so the problem never occurs, as the client makes sure all necessary form fields are filled appropriately. And of course, this is what I would also do in a real application. But I usually disable it during development to check whether the server side validation works as expected, and if I disable that in a `make:crud`application, it crashed in exactly  `form->handleRequest()`.

**Reply by ahundiak** [reply]
Might want to open an issue on the maker bundle or perhaps search the existing issues to see if this has been brought up before.  The command certainly has enough information to set a default value for non-nullable strings and integers.  But the author might have their reasons.

**Reply by aszenz** [reply]
If I understood your problem clearly, you want to prevent updating a field to null/empty string with a nice error message. 

Have you tried using `setter` form field property, I find it useful to pre validate things before calling the actual entity setter, you could throw a form error their if the value is null.

https://symfony.com/blog/new-in-symfony-5-2-form-mapping-callbacks

When using forms with entities I always try to use `getter` and  `setter` for each field, it helps in decoupling things a bit, combine this with `empty_data` to construct a valid entity first before submitting form values, u can make sure that your entity code is properly encapsulated.

**Comment by ahundiak** [comment, marked as answer]
It is one of those age old discussions that will never be settled.  Kind of comforting in a way.

Doctrine is database driven with one to one mappings between Doctrine entities and database tables.  In principle your business entities should be database ignorant and designed strictly via business requirements.  The `problem` is that for a large portions of the applications I work on, Doctrine and business entities are the same.  They are basically CRUD.  So why bother introducing complexity that is not needed.

On the other hand the fun stuff is where you actually have business entities that truly are database independent.  It is typically worth the effort to have form specific DTO because the mapping is no longer just one to one.  When it comes time to persist you really do need some custom logic to determine what data goes where.
